fix: do not accept tx with unexpected pubkeys.

AnExsomnis opened the pull request:
Closes: #5431

AnNiran committed:
fix: do not accept tx with unexpected pubkeys.

wwestgarth commented on processor/abci.go:
```diff
@@ -746,6 +747,12 @@ func (app *App) OnCheckTx(ctx context.Context, _ tmtypes.RequestCheckTx, tx abci
 		app.log.Debug("entering checkTx", logging.String("tid", tx.GetPoWTID()), logging.String("command", tx.Command().String()))
 	}
 
+	if !app.exec.MarketsExpectKey(tx.PubKeyHex()) {
```
I think we only want to do this check if the incoming transaction is a `SubmitOracleDataCommand`?

AnExsomnis replied:
I have corrected this here https://github.com/vegaprotocol/vega/pull/5534/commits/8ed6a685a32b5cdaf527f4bd083f72c17ff2a55a#diff-43cb75f5e3d4bfc8b8fc4540853a32c32346ec1fb151bbaed720d7bf07a4a139R822-R825
Thanks!

wwestgarth commented on execution/engine.go:
```diff
@@ -273,6 +274,19 @@ func (e *Engine) IsEligibleForProposerBonus(marketID string, value *num.Uint) bo
+func (e *Engine) MarketsExpectKey(key string) bool {
+	for _, mkt := range e.markets {
+		for k := range mkt.mkt.TradableInstrument.Instrument.GetFuture().OraclesPubKeys() {
+			if strings.Compare(k, key) {
```
I'm probably missing something but whats wrong with `if k == key`?

AnExsomnis replied:
I have added the logic here https://github.com/vegaprotocol/vega/pull/5534/commits/8ed6a685a32b5cdaf527f4bd083f72c17ff2a55a#diff-175c28f9d4bf1098cfc45b2c0071f08ea70dc56c117b273fee2ef110a876487eR15 it to aim for a more efficient comparison:  fail faster or/and in the first not matching rune, instead of traversing the whole strings.

ValentinTrinque commented on execution/engine.go:
```diff
@@ -273,6 +274,19 @@ func (e *Engine) IsEligibleForProposerBonus(marketID string, value *num.Uint) bo
 }
 
+// MarketsExpectKey checks if the currently existing markets use the provided key in their subscriptions.
+func (e *Engine) MarketsExpectKey(key string) bool {
```
Based on my previous comments, the algorithm is a bit more complex than this. 

We consider the OracleData to match an OracleSpec when all the public keys that signed the oracle data are defined in the oracle spec.

So getting a **list** of the _watched public keys_ won't be enough. You will need a more complex data structure.

For example:
* Market A listens to 0x11111
* Market B listens to 0x22222
* An oracle data signed by 0x11111 will match
* But an oracle data signed by 0x11111 and 0x22222 won't match, because no one is listening at **both keys at the same time**

AnExsomnis replied:
Yes, this has been addressed here https://github.com/vegaprotocol/vega/pull/5534#issuecomment-1177353456
Thanks!

AnExsomnis commented on processor/abci.go:
```diff
@@ -48,6 +48,7 @@ import (
 
 var (
 	ErrPublicKeyCannotSubmitTransactionWithNoBalance  = errors.New("public key cannot submit transaction without balance")
+	ErrUnexpectedTxPubKey                             = errors.New("received unexpected public key in transaction")
```
Yes, makes sense; Thanks!

AnExsomnis commented:
@ValentinTrinque ,
I have reworked the logic in a new commit (for now) - https://github.com/vegaprotocol/vega/pull/5534/commits/a5d8d5bcd4180e37b763161986d92e913680b01b
The check for existing active keys is transferred to the OracleSpecs and is used by the OracleEngine (as entities that are aware of all keys and should handle this).
Also, I have added this method call inside processor/abci.go - CheckSubmitOracleData as this check could simply be done inside the handler instead of having same call for normalization somewhere else.
Please take a look when you have time and let's discuss.

AnExsomnis commented on processor/abci.go:
```diff
@@ -818,6 +819,10 @@ func (app *App) canSubmitTx(tx abci.Tx) (err error) {
 			}
 		}
+	case txn.SubmitOracleDataCommand:
+		if !app.exec.MarketsExpectKey(tx.PubKeyHex()) {
```
Thank you, this has been addressed here https://github.com/vegaprotocol/vega/pull/5534#issuecomment-1177353456

AnExsomnis commented on types/market.go:
```diff
@@ -409,6 +409,21 @@ func (f Future) String() string {
+
+	return uKeys
+}
+
```
Yes, this has been addressed here https://github.com/vegaprotocol/vega/pull/5534#issuecomment-1177353456
Thanks!

ValentinTrinque commented on processor/abci.go:
```diff
@@ -1244,9 +1244,12 @@ func (app *App) CheckSubmitOracleData(_ context.Context, tx abci.Tx) error {
-	_, err := app.oracles.Adaptors.Normalise(pubKey, *data)
+	oracleData, err := app.oracles.Adaptors.Normalise(pubKey, *data)
+	if err != nil {
+		return err
```
Except if the underlaying errror is very clear for what's happening, could you wrap this error with `fmt.Errof()` ?

Something that can help understand *where* things went wrong :)

AnExsomnis replied:
Yes,
I did some optimizations on this here https://github.com/vegaprotocol/vega/pull/5534/commits/587bd850dcca9027fd5422ef1cf618ff087a449d#diff-43cb75f5e3d4bfc8b8fc4540853a32c32346ec1fb151bbaed720d7bf07a4a139R59

ValentinTrinque commented on oracles/oracle_spec.go:
```diff
@@ -144,6 +144,12 @@ func isInternalOracleData(data OracleData) bool {
 
+// matchPubKeys tries to match the public keys from the provided OracleData object with the ones
+// present in the Spec.
+func (s *OracleSpec) matchPubKeys(data OracleData) bool {
```
Since it's a method called by an external component, I think we should make this method pubic as it's now part of the API of the oracle spec. It states the intention better.

AnExsomnis replied:
All right, we logically separate them, I understand.
Thanks!

AnExsomnis replied:
This has been corrected here https://github.com/vegaprotocol/vega/pull/5534/commits/587bd850dcca9027fd5422ef1cf618ff087a449d#diff-c31fe3f1776ec18d9a8482d6f1b00732f84023362e6183a17e18e379b0b28ebcR147-R152

ValentinTrinque commented on oracles/engine.go:
```diff
@@ -64,6 +64,19 @@ func NewEngine(
 
+// ListensToPubKeys checks if the public keys from provided OracleData are among the keys
+// current OracleSpecs listen to.
+func (e *Engine) ListensToPubKeys(data OracleData) error {
+	_, err := e.subscriptions.filterSubscribers(func(spec OracleSpec) (bool, error) {
+		if !spec.matchPubKeys(data) {
+			return false, fmt.Errorf("provided data public keys do not match currently active in specs")
+		}
+		return true, nil
+	})
+
+	return err
+}
+
```
Overall, this is very good. I think you can optimize this by implementing a `hasAnySubscribers() bool` method on the subscriptions structure, so it doesn't iterate over all the specs. We just needs to know if _at least one_ matches.

ValentinTrinque commented on oracles/engine.go:
```diff
@@ -64,6 +64,19 @@ func NewEngine(
 
+// ListensToPubKeys checks if the public keys from provided OracleData are among the keys
+// current OracleSpecs listen to.
+func (e *Engine) ListensToPubKeys(data OracleData) error {
```
And to finish, could you add some tests on this ?

ValentinTrinque commented on processor/abci.go:
```diff
@@ -48,13 +48,17 @@ import (
 	ErrCheckpointRestoreDisabledDuringBootstrap       = errors.New("checkpoint restore disabled during bootstrap")
 	ErrAwaitingCheckpointRestore                      = errors.New("transactions not allowed while waiting for checkpoint restore")
+	ErrOracleDataNormalization                        = func(err error) error {
+		return fmt.Errorf("error normalizing incoming oracle data: %s", err.Error())
```
You can use `%w`, it wraps the underlaying error and works better with error API `errors.Is()`

```suggestion
		return fmt.Errorf("error normalizing incoming oracle data: %w", err)
```

wwestgarth commented on oracles/engine.go:
```diff
@@ -64,6 +64,14 @@ func NewEngine(
+// ListensToPubKeys checks if the public keys from provided OracleData are among the keys
+// current OracleSpecs listen to.
+func (e *Engine) ListensToPubKeys(data OracleData) bool {
+	return e.subscriptions.hasAnySubscribers(func(spec OracleSpec) bool {
```
I'm a bit late now its merged (sorry), but given `CheckTx` happens concurrently with the block processing I think we're going to need a lock on the subscribers. We could be adding a new subscriber (due to a new market creation) while at the same time an oracle-command is being checked.
